src/internal/coredb/projects.go

```diff
@@ -0,0 +1,176 @@
+	}
+	*project = pfs.ProjectInfo{
+		Project:     &pfs.Project{Name: row.Name},
+		Description: row.Description,
+		CreatedAt:   projectTimestamp}
+	iter.index++
+	return nil
+}
+
+// ListProject returns a ProjectIterator that exposes a Next() function for retrieving *pfs.ProjectInfo references.
+func ListProject(ctx context.Context, db *pachsql.DB) (*ProjectIterator, error) {
+	page, err := listProjectPage(ctx, db, 100, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list projects")
+	}
+	iter := &ProjectIterator{
+		projects: page,
+		db:       db,
+	}
+	return iter, nil
+}
+
+func listProjectPage(ctx context.Context, db *pachsql.DB, pageSize, pageNum int) ([]projectRow, error) {
+	var page []projectRow
+	if err := db.SelectContext(ctx, &page, "SELECT name,description,created_at FROM core.projects ORDER BY id ASC LIMIT $1 OFFSET $2",
+		pageSize, pageSize*pageNum); err != nil {
+		return nil, errors.Wrap(err, "could not get project page")
+
+	}
+	return page, nil
+}
+
+// QueryExecer defines an interface for functions shared across sqlx.Tx and sqlx.DB types.
+// Functions that take a querier support both running inside a transaction and outside it.
+type QueryExecer interface {
+	sqlx.QueryerContext
+	sqlx.ExecerContext
+}
+
+// CreateProject creates an entry in the core.projects table.
+func CreateProject(ctx context.Context, queryExecer QueryExecer, project *pfs.ProjectInfo) error {
+	_, err := queryExecer.ExecContext(ctx, "INSERT INTO core.projects (name, description) VALUES ($1, $2);", project.Project.Name, project.Description)
+	//todo: insert project.authInfo into auth table.
+	return errors.Wrap(err, "failed to create project")
+}
+
+// DeleteProject deletes an entry in the core.projects table.
+func DeleteProject(ctx context.Context, queryExecer QueryExecer, projectName string) error {
+	result, err := queryExecer.ExecContext(ctx, "DELETE FROM core.projects WHERE name = $1;", projectName)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete project")
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return errors.Wrap(err, "could not get affected rows")
+	}
+	if rowsAffected == 0 {
+		return errors.Wrap(fmt.Errorf("project %s does not exist", projectName), "failed to delete project")
+	}
+	return nil
+}
+
+func DeleteAllProjects(ctx context.Context, queryExecer QueryExecer) error {
+	_, err := queryExecer.ExecContext(ctx, "TRUNCATE core.projects;")
+	return errors.Wrap(err, "could not delete all project rows")
+}
+
+// GetProject retrieves an entry from the core.projects table by project name.
+func GetProject(ctx context.Context, queryExecer QueryExecer, projectName string) (*pfs.ProjectInfo, error) {
+	return getProject(ctx, queryExecer, "name", projectName)
+}
+
+// GetProjectByID is like GetProject, but retrieves an entry using the row id.
+func GetProjectByID(ctx context.Context, queryExecer QueryExecer, id int64) (*pfs.ProjectInfo, error) {
```

Comment:
IMO: `Get_By_` suggests the use of an index, or something other than the identity and `Get_` suggests retrieval by the identity (whatever that is).

Reply:
ID is an index by nature of being a primary key, right? 

Having said that, I'm not sure we will use this function?  I feel like the IDs are always going to be for joins, whereas getting things by name is how every RPC is going to work.  Remember, IDs never leave the system boundary; the API doesn't know them.

Reply:
You can have 1 `Get_` method.  You can have as many `Get_By_` methods as you want.  You get 1 primary key, you can have as many indexes as you want.

Reply:
Yeah, it seems like the names are inverted. `GetProject` should take the id and `GetProjectByName` should be used for the name. Internal usage may generally lean towards `Get*` with an id, specifically for the background processes that are watching and will receive ids. I don't think we have any background processes for projects yet, but repos, branches, commits, etc. will be using ids.

Reply:
Based on our conversation, we want to standardize on retrieval by ID in our codebase. GetProject will be based on ID and we will have a GetProjectByName for RPCs.